metagraph/src/annotation/taxonomy/tax_classifier.cpp

```diff
@@ -0,0 +1,300 @@
+}
+
+void TaxonomyClsAnno::read_tree(const std::string &tax_tree_filepath, ChildrenList *tree) {
+    std::ifstream f(tax_tree_filepath);
+    if (!f.good()) {
+        logger->error("Error: Failed to open Taxonomic Tree file {}", tax_tree_filepath);
+        exit(1);
+    }
+
+    std::string line;
+    tsl::hopscotch_map<TaxId, TaxId> full_parents_list;
+    while (getline(f, line)) {
+        if (line == "") {
+            logger->error("Error: The Taxonomic Tree file contains empty lines. "
+                          "Please make sure that this file was not manually modified: {}",
+                          tax_tree_filepath);
+            exit(1);
+        }
+        std::vector<std::string> parts = utils::split_string(line, "\t");
+        if (parts.size() <= 2) {
+            logger->error("Error: The Taxonomic tree filepath contains incomplete lines. "
+                          "Please make sure that this file was not manually modified: {}",
+                          tax_tree_filepath);
+            exit(1);
+        }
+        uint32_t act = std::stoul(parts[0]);
+        uint32_t parent = std::stoul(parts[2]);
+        full_parents_list[act] = parent;
+        node_parent_[act] = parent;
+    }
+
+    std::vector<TaxId> relevant_taxids;
+    // 'considered_relevant_taxids' is used to make sure that there are no duplications in 'relevant_taxids'.
+    tsl::hopscotch_set<TaxId> considered_relevant_taxids;
+
+    if (accversion_to_taxid_map_.size()) {
+        // Store only the taxonomic nodes that exists in the annotation matrix.
+        for (const std::pair<std::string, TaxId> &pair : accversion_to_taxid_map_) {
+            relevant_taxids.push_back(pair.second);
+            considered_relevant_taxids.insert(pair.second);
+        }
+    } else {
+        // If 'this->accversion_to_taxid_map' is empty, store the entire taxonomic tree.
+        for (auto it : full_parents_list) {
+            relevant_taxids.push_back(it.first);
+            considered_relevant_taxids.insert(it.first);
+        }
+    }
+    assert(relevant_taxids.size());
+
+    uint64_t num_taxid_failed = 0; // num_taxid_failed is used for logging only.
+    for (uint32_t i = 0; i < relevant_taxids.size(); ++i) {
+        const TaxId taxid = relevant_taxids[i];
+        if (!full_parents_list.count(taxid)) {
+            num_taxid_failed += 1;
+            continue;
+        }
+
+        if (not considered_relevant_taxids.count(full_parents_list[taxid])) {
+            relevant_taxids.push_back(full_parents_list[taxid]);
+            considered_relevant_taxids.insert(full_parents_list[taxid]);
+        }
+
+        // Check if the current taxid is the root.
+        if (taxid == full_parents_list[taxid]) {
+            root_node_ = taxid;
+        }
+    }
+    if (num_taxid_failed) {
+        logger->warn("During the tax_tree_filepath {} parsing, {} taxids were not found out of {} total evaluations.",
```

Comment:
```suggestion
        logger->warn("During the tax_tree_filepath {} parsing, {} taxids were not found out of {} total evaluations",
```

Reply:
Regarding logger messages, I am not very sure in what context to add the period and when to remove it. Can you please list some small examples to understand how to write these in the future?